### PR TITLE
[SPARK-31164][SQL] Inconsistent rdd and output partitioning for bucket table when output doesn't contain all bucket columns

### PR DESCRIPTION
**Pull request by wzhfy**
### What changes were proposed in this pull request?

For a bucketed table, when deciding output partitioning, if the output doesn't contain all bucket columns, the result is `UnknownPartitioning`. But when generating rdd, current Spark uses `createBucketedReadRDD` because it doesn't check if the output contains all bucket columns. So the rdd and its output partitioning are inconsistent.

### Why are the changes needed?

To fix a bug.

### Does this PR introduce any user-facing change?

No.

### How was this patch tested?

Modified existing tests.

### CODE REVIEW

**Review by cloud-fan — APPROVED** [review]
good catch!

File: sql/core/src/main/scala/org/apache/spark/sql/execution/DataSourceScanExec.scala [hunk 1]
```diff
@@ -340,6 +341,7 @@ case class FileSourceScanExec(
         "PartitionFilters" -> seqToString(partitionFilters),
         "PushedFilters" -> seqToString(pushedDownFilters),
         "DataFilters" -> seqToString(dataFilters),
+        "BucketedScan" -> bucketedScan.toString,
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/DataSourceScanExec.scala]: this breaks some explain tests, shall we do it later?

Comment (wzhfy) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/DataSourceScanExec.scala]: ok

**Comment by dongjoon-hyun** [comment]
cc @dbtsai since this is related to table bucketing.

File: sql/core/src/test/scala/org/apache/spark/sql/sources/BucketedReadSuite.scala [hunk 2]
```diff
@@ -100,6 +100,12 @@ abstract class BucketedReadSuite extends QueryTest with SQLTestUtils {
     }
   }
 
+  private def getFilScan(plan: SparkPlan): FileSourceScanExec = {
```
Comment (dongjoon-hyun) [inline sql/core/src/test/scala/org/apache/spark/sql/sources/BucketedReadSuite.scala]: `getFilScan` -> `getFileScan`?

**Review by dongjoon-hyun — APPROVED** [review]
+1, LGTM (except one typo issue).

**Comment by wzhfy** [comment]
thanks for reviewing, merged to master/3.0

**Comment by dongjoon-hyun** [comment]
Thank you, @wzhfy and @cloud-fan

**Comment by manuzhang** [comment]
@cloud-fan @wzhfy 
I'm wondering whether bucketed scan with `UnknownPartitioning` is valid.

As I understand it, bucketed scan has two effects
1. decides how we partition input files
2. benefits downstream operators (e.g. bucket join)

When 2 is not effective, we may still make use of 1. For example, we can have a bounded number of `FilePartition`, which could blow up driver memory without bucketed scan. Or is there a correctness issue I've missed?

**Comment by cloud-fan** [comment]
I don't think limiting the number of file partitions was the design goal of the bucketed table. We can set `spark.sql.files.maxPartitionBytes` to a large number like `1GB` to reduce the partitions.

**Comment by manuzhang** [comment]
Yes, but the effect has leaked into user space and this breaks it. As to my original question, is `HashPartitioning` a hard requirement for bucketed scan?

**Comment by cloud-fan** [comment]
I don't think so, and Spark will disable bucketed scan if it has no benefit for downstream operators, see the rule `DisableUnnecessaryBucketedScan`

**Comment by manuzhang** [comment]
That rule can be disabled but this change can't.

**Comment by cloud-fan** [comment]
That rule means this is by design. We believe bucketed scan is more expensive than a normal scan and only want to use it if it can avoid shuffles. Maybe this does not apply in your case but it applies in many other cases. If you found this issue before we release it, we can revert to avoid perf regression. But revert is not an option today as it may cause perf regression for more people.

I'd suggest revisiting your requirement and considering `spark.sql.files.maxPartitionBytes`. We should not abuse bucketed table here.

**Comment by manuzhang** [comment]
I'm not asking for revert here but to explore an option to disable this behavior, and hence the question about partitioning and bucketed scan.
Bucket table is built to avoid shuffle but it's a table and we cannot prevent downstream users from using it in other ways. As a platform, we'd better keep backward compatibility with some general options rather than asking users to fine tune their jobs.

**Comment by cloud-fan** [comment]
I'm fine to keep backward compatible with some "by-accident" features if the cost is small. Feel free to open a PR to bring back the old behavior if
1. we do not re-introduce the correctness bug
2. the code is simple (less maintenance cost)

**Comment by manuzhang** [comment]
@cloud-fan please help check https://github.com/apache/spark/pull/36733

**Pull request closed without merging**